docs/docs/reference/gen_notebooks/hallucination.md

```diff
@@ -0,0 +1,450 @@
+    model_name: str = "gpt-3.5-turbo"
+
+    @weave.op()
+    def predict(self, question: str) -> Dict[str, Any]:
+        '''
+        Generates an answer to the question based on retrieved context.
+        Returns a dict with 'answer', 'context', and 'model_name'.
+        '''
+        # Retrieve context
+        context = retrieve_context(question)
+        # Generate answer
+        answer = generate_answer(question, context, self.model_name)
+        return {'answer': answer, 'context': context, 'model_name': self.model_name}
+
+```
+
+## 📝 Define Scorer Function
+
+We'll define a scorer function that computes the faithfulness score of the model's answer. This function will be used by Weave's `Evaluation` class.
+
+
+```python
+
+@weave.op()
+def hallucination_scorer(model_output: Dict[str, Any]) -> Dict[str, Any]:
+    '''
+    Scorer function that computes the factual score of the model's answer for evaluating hallucination.
+    '''
+    answer = model_output['answer']
+    context = model_output['context']
+    model_name = model_output['model_name']
+    statements = break_down_answer_into_statements(answer, model_name)
+    total_statements = len(statements)
+    factual_statements = 0
+    statement_results = []
+    for statement in statements:
+        result = check_statement_hallucination(context, answer)
+        factual_statements += result['prediction']
+        statement_results.append({
+            'prediction': result['prediction'],
+            'statement': result['statement']
+        })
+    # Calculate faithfulness score
+    if total_statements > 0:
+        factual_score = factual_statements / total_statements
+    else:
+        factual_score = 0
+    # Return results
+    return {
+        'factual_score': factual_score,
+        'statement_results': statement_results,
+    }
+
+```
+
+## 🚀 Run Evaluation Using Weave's `Evaluation` Class
+
+We'll use Weave's `Evaluation` class to run the evaluation, ensuring that the results are stored in the **'eval'** section of Weave.
+
+
+```python
+# Import Weave's Evaluation class
+from weave import Evaluation
+
+# Initialize Weave
+weave.init(PROJECT)
+
+# Apply nest_asyncio to allow nested event loops in Colab
+import nest_asyncio
+nest_asyncio.apply()
+
+# Run the evaluation for both models
+import asyncio
+
+# Define the models to evaluate
+model_names = ["gpt-3.5-turbo", "gpt-4o"]
+
+for model_name in model_names:
+    print(f"Running evaluation with model: {model_name}")
+    # Instantiate the evaluator model with the specified model name
+    evaluator_model = HallucinationEvaluator(model_name=model_name)
+
+    # Define the evaluation
+    evaluation = Evaluation(
+        dataset=dataset,  # the dataset we have defined earlier
+        scorers=[hallucination_scorer],  # the scorer function
+    )
+
+    # Run the evaluation
+    summary = asyncio.run(evaluation.evaluate(evaluator_model))
+
+    print(f"Completed evaluation with model: {model_name}\n")
+
```

Comment:
Would be good to show some results here, like we see the crappier model hallucinates more or something.

Reply:
Let's add a caption to the image explaining what this is telling us.  Something like in our example 3.5 turbo had a factual score of 1 while gpt-4o got 0.8 which means 4o is hallucinating more which doesn't make any sense 🤣

Reply:
Yes it sounds surprising, but here is the reason observed( I ran all the code with more changes to make it more observable in weave and still get the response for GPT3.5 better ). I added this to notebooks too:

_During the evaluation for this particular usecase, it was observed that GPT-4o tends to produce more hallucinations than GPT-3.5, which may seem unexpected at first. However, this difference can be explained by how hallucinations manifest in Retrieval-Augmented Generation (RAG).


A common type of hallucination in RAG is factual but ungrounded. For example, given the premise "The capital of France is Berlin," a generated response stating "The capital of France is Paris" would be considered a hallucination—even though it is factually correct—because the generated statement is not derived from the provided retrieval context.

This issue arises when the model generates answers that are based on its pre-trained knowledge rather than the specific textual data retrieved as part of the RAG process.

GPT-4o, with its tendency to produce more detailed and precise responses, sometimes is more prone to this type of hallucination. As its answers are often longer and more comprehensive, they contain remote facts or statements that may not align with or be verifiable from the retrieved context. As a result, the evaluation score for GPT-4 may appear lower, even though some of the hallucinated statements might still be factually accurate._